Segfault in 1.8.x and 1.7.5 with Consul Connect

I've just upgraded a test Kubernetes cluster to 1.8.1 from 1.7.1 and my agent instances are crashing over and over with a segfault

I'm not sure what call is triggering this, there are a number of services in the cluster using Consul Connect but without the consul agents running its hard to isolate the cause of this crash as... well everything is crashing and restarting as there's no agent available.

```
    2020-08-03T10:24:21.119Z [INFO]  agent: Join cluster completed. Synced with initial agents: cluster=LAN num_agents=3
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x24b8dfc]

goroutine 131 [running]:
github.com/hashicorp/consul/agent/xds.(*Server).process(0xc00012a840, 0x37e3240, 0xc000656b70, 0xc000840660, 0x0, 0x0)
	/home/circleci/project/consul/agent/xds/server.go:334 +0x65c
github.com/hashicorp/consul/agent/xds.(*Server).StreamAggregatedResources(0xc00012a840, 0x37e3240, 0xc000656b70, 0x3781900, 0xc00012a840)
	/home/circleci/project/consul/agent/xds/server.go:171 +0xc5
github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2._AggregatedDiscoveryService_StreamAggregatedResources_Handler(0x30f5040, 0xc00012a840, 0x37d91c0, 0xc0008a2180, 0x53c8e90, 0xc000733a00)
	/go/pkg/mod/github.com/envoyproxy/go-control-plane@v0.9.5/envoy/service/discovery/v2/ads.pb.go:197 +0xad
google.golang.org/grpc.(*Server).processStreamingRPC(0xc000375380, 0x37e9240, 0xc000afa600, 0xc000733a00, 0xc000535e00, 0x5365e80, 0x0, 0x0, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.25.1/server.go:1211 +0xd1e
google.golang.org/grpc.(*Server).handleStream(0xc000375380, 0x37e9240, 0xc000afa600, 0xc000733a00, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.25.1/server.go:1291 +0xcd6
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc000857240, 0xc000375380, 0x37e9240, 0xc000afa600, 0xc000733a00)
	/go/pkg/mod/google.golang.org/grpc@v1.25.1/server.go:722 +0xa1
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/pkg/mod/google.golang.org/grpc@v1.25.1/server.go:720 +0xa1

```

edit:

Managed to pin this down a bit more.  
I can reliably reproduce the crash when there is any Consul Connect enabled service running (e.g. an Envoy proxy operating as a sidecar pulling config from the local agent via xDS) and the agent restarts.  
I'm using consul-helm to setup my agents and the connect injector to setup Envoy sidecars.

* Start consul agent
* Start consul connect enabled service
* Everything is fine
* Restart consul agent
* Agent segfaults when Envoy makes a gRPC call

I've tested this with Consul 1.7.1, 1.7.3, 1.7.4, 1.7.5 and 1.8.1 with the latest [supported](https://www.consul.io/docs/connect/proxies/envoy#supported-versions) Envoy version for each (so 1.13.1 and 1.14.2, specifically the `envoyproxy/envoy-alpine` image variant)

Only 1.7.5 and 1.8.1 crash when the agent is restarted. 1.7.4 is fine.

The only commit I can see between 1.7.4 and 1.7.5 that looks relevant is https://github.com/hashicorp/consul/pull/8266

On Consul versions that don't crash, and on versions that do crash which are already running when the Connect service starts, I do see a log entry. Not sure if this is relevant though, the consul-helm [changelog](https://github.com/hashicorp/consul-helm/blob/master/CHANGELOG.md#0220-june-18-2020) has a note about an incorrect health check that will fail on 1.8.0, but i'm using consul-k8s:0.18.0 for all these tests. 

```
2020-08-03T16:30:50.972Z [ERROR] agent.proxycfg: watch error: id=service-http-checks:api-gateway-7f4f97b79f-dhsqm-api-gateway error="invalid type for service checks response: cache.FetchResult, want: []structs.CheckType"
```

--- a/command/connect/envoy/bootstrap_tpl.go
+++ b/command/connect/envoy/bootstrap_tpl.go
@@ -188,7 +188,6 @@ const bootstrapTemplate = `{
     "cds_config": { "ads": {} },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/CONSUL_HTTP_ADDR-with-https-scheme-enables-tls.golden
+++ b/command/connect/envoy/testdata/CONSUL_HTTP_ADDR-with-https-scheme-enables-tls.golden
@@ -97,7 +97,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/access-log-path.golden
+++ b/command/connect/envoy/testdata/access-log-path.golden
@@ -88,7 +88,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/defaults.golden
+++ b/command/connect/envoy/testdata/defaults.golden
@@ -88,7 +88,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/existing-ca-file.golden
+++ b/command/connect/envoy/testdata/existing-ca-file.golden
@@ -97,7 +97,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/extra_-multiple.golden
+++ b/command/connect/envoy/testdata/extra_-multiple.golden
@@ -110,7 +110,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/extra_-single.golden
+++ b/command/connect/envoy/testdata/extra_-single.golden
@@ -101,7 +101,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/grpc-addr-config.golden
+++ b/command/connect/envoy/testdata/grpc-addr-config.golden
@@ -88,7 +88,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/grpc-addr-env.golden
+++ b/command/connect/envoy/testdata/grpc-addr-env.golden
@@ -88,7 +88,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/grpc-addr-flag.golden
+++ b/command/connect/envoy/testdata/grpc-addr-flag.golden
@@ -88,7 +88,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/grpc-addr-unix.golden
+++ b/command/connect/envoy/testdata/grpc-addr-unix.golden
@@ -87,7 +87,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/ingress-gateway-address-specified.golden
+++ b/command/connect/envoy/testdata/ingress-gateway-address-specified.golden
@@ -161,7 +161,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/ingress-gateway-no-auto-register.golden
+++ b/command/connect/envoy/testdata/ingress-gateway-no-auto-register.golden
@@ -161,7 +161,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/ingress-gateway-register-with-service-and-proxy-id.golden
+++ b/command/connect/envoy/testdata/ingress-gateway-register-with-service-and-proxy-id.golden
@@ -161,7 +161,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/ingress-gateway-register-with-service-without-proxy-id.golden
+++ b/command/connect/envoy/testdata/ingress-gateway-register-with-service-without-proxy-id.golden
@@ -161,7 +161,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/ingress-gateway.golden
+++ b/command/connect/envoy/testdata/ingress-gateway.golden
@@ -161,7 +161,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/stats-config-override.golden
+++ b/command/connect/envoy/testdata/stats-config-override.golden
@@ -46,7 +46,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/token-arg.golden
+++ b/command/connect/envoy/testdata/token-arg.golden
@@ -88,7 +88,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/token-env.golden
+++ b/command/connect/envoy/testdata/token-env.golden
@@ -88,7 +88,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/token-file-arg.golden
+++ b/command/connect/envoy/testdata/token-file-arg.golden
@@ -88,7 +88,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/token-file-env.golden
+++ b/command/connect/envoy/testdata/token-file-env.golden
@@ -88,7 +88,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

--- a/command/connect/envoy/testdata/zipkin-tracing-config.golden
+++ b/command/connect/envoy/testdata/zipkin-tracing-config.golden
@@ -121,7 +121,6 @@
     },
     "ads_config": {
       "api_type": "GRPC",
-      "set_node_on_first_message_only": true,
       "grpc_services": {
         "initial_metadata": [
           {

